File: Spiral-Matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
function spiralOrder(matrix) {
    const res = [];

    if (!matrix || matrix.length === 0 || matrix[0].length === 0) return res;

    function spiral(matrix, nrow, ncol, res, row, col, nxtrow, nxtcol) {
        if (nrow === 0 || ncol === 0) return;

        for (let i = 0; i < ncol; i++) {
            row += nxtrow;
            col += nxtcol;
            res.push(matrix[row][col]);
        }

        spiral(matrix, ncol, nrow - 1, res, row, col, nxtcol, -nxtrow);
    }

    spiral(matrix, matrix.length, matrix[0].length, res, 0, -1, 0, 1);
    return res;
}

File: Spiral-Matrix.cpp
class Solution {
private:
    void spiral(vector<vector<int>>& matrix, int nrow, int ncol,
                vector<int>& res, int row, int col, int nxtrow, int nxtcol) {
        if (nrow == 0 || ncol == 0) return;

        for (int i = 0; i < ncol; ++i) {
            row = row + nxtrow;
            col = col + nxtcol;
            res.push_back(matrix[row][col]);
        }
        spiral(matrix, ncol, nrow - 1, res, row, col, nxtcol, -nxtrow);
    }

public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;

        if (matrix.empty() || matrix[0].empty()) return res;

        spiral(matrix, matrix.size(), matrix[0].size(), res, 0, -1, 0, 1);
        return res;
    }
};
